src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,678 @@
+#include "librbd/AioCompletion.h"
+#include "librbd/Journal.h"
+#include "librbd/MirroringWatcher.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::CreateRequest: "
+
+namespace librbd {
+namespace image {
+
+using util::create_context_callback;
+using util::create_rados_ack_callback;
+
+namespace {
+
+bool validate_features(CephContext *cct, uint64_t features, bool force_non_primary) {
+  if ((features & RBD_FEATURE_FAST_DIFF) != 0 &&
+      (features & RBD_FEATURE_OBJECT_MAP) == 0) {
+    lderr(cct) << "cannot use fast diff without object map" << dendl;
+    return false;
+  }
+  if ((features & RBD_FEATURE_OBJECT_MAP) != 0 &&
+      (features & RBD_FEATURE_EXCLUSIVE_LOCK) == 0) {
+    lderr(cct) << "cannot use object map without exclusive lock" << dendl;
+    return false;
+  }
+  if ((features & RBD_FEATURE_JOURNALING) != 0) {
+    if ((features & RBD_FEATURE_EXCLUSIVE_LOCK) == 0) {
+      lderr(cct) << "cannot use journaling without exclusive lock" << dendl;
+      return false;
+    }
+  } else if (force_non_primary) {
+    assert(false);
+  }
+
+  return true;
+}
+
+bool validate_layout(CephContext *cct, uint64_t size, file_layout_t &layout) {
+  if (!librbd::ObjectMap::is_compatible(layout, size)) {
+    lderr(cct) << "image size not compatible with object map" << dendl;
+    return false;
+  }
+
+  return true;
+}
+
+} // anonymous namespace
+
+template<typename I>
+CreateRequest<I>::CreateRequest(I *image_ctx, uint64_t size, int order, uint64_t features,
+                                uint64_t stripe_unit, uint64_t stripe_count,
+                                uint8_t journal_order, uint8_t journal_splay_width,
+                                const std::string &journal_pool,
+                                const std::string &non_primary_global_image_id,
+                                const std::string &primary_mirror_uuid,
+                                Context *on_finish) :
+  m_image_ctx(image_ctx), m_size(size), m_order(order), m_features(features),
+  m_stripe_unit(stripe_unit), m_stripe_count(stripe_count),
+  m_journal_order(journal_order), m_journal_splay_width(journal_splay_width),
+  m_journal_pool(journal_pool), m_non_primary_global_image_id(non_primary_global_image_id),
+  m_primary_mirror_uuid(primary_mirror_uuid), m_on_finish(on_finish) {
+
+  m_cct = image_ctx->cct;
+
+  m_id_obj = util::id_obj_name(image_ctx->name);
+  m_header_obj = util::header_name(image_ctx->id);
+  m_objmap_name = ObjectMap::object_map_name(image_ctx->id, CEPH_NOSNAP);
+
+  m_layout = file_layout_t();
+  m_layout.object_size = 1ull << m_order;
+  if (m_stripe_unit == 0 || m_stripe_count == 0) {
+    m_layout.stripe_unit = m_layout.object_size;
+    m_layout.stripe_count = 1;
+  } else {
+    m_layout.stripe_unit = m_stripe_unit;
+    m_layout.stripe_count = m_stripe_count;
+  }
+
+  m_force_non_primary = !non_primary_global_image_id.empty();
+}
+
+// TODO: dtor()
```

Comment:
That's a left over from previous patch. I haven't incorporated all earlier comments -- sent this out so as get a successful build run and the last patchset has unittest_librbd failing..